### ISSUE TITLE
Divide by zero when passing in omega with 0 on the diagonal

### ISSUE THREAD

**Issue by jacobms** [robertmartin8/PyPortfolioOpt#70]
I'm slowly but surely trying to recreate the results in the Idzorek paper using PyPortfolioOpt.

On p. 21 he mentions _"Setting all of the diagonal elements of omega equal to zero is equivalent to specifying 100% confidence in all of the K views"_.

But when I pass in a zero diagonal matrix as omega I get a divide by zero error when I call bl_returns(). This is also a problem with bl_cov()

During the Coursera course https://www.coursera.org/learn/advanced-portfolio-construction-python the instructor notes that inverting omega is not always possible so he references two alternative ways to calculate the expected returns and covariance matrix:

![image](https://user-images.githubusercontent.com/190806/74397794-cedb9c00-4e15-11ea-8cd9-137937bd1128.png)

Maybe that can inspire?

The instructor specifically refers to https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1314585 where Walters has rewritten the two expressions to versions without inversing omega. When I compare the formulas on the screen dump with the Walters paper it seems that the first plus should be a minus but given that it is quite a long time since I deep dived into this kind of math I might be wrong.

BTW, I can approximate a solution by filling the diagonal with np.fill_diagonal(omega_100, 1E-8) but that seems pretty hack-ish.

https://github.com/robertmartin8/PyPortfolioOpt/blob/bdbc9fec0f192edcc42abfcd95ecfc07e3dfdab3/pypfopt/black_litterman.py#L273

**Comment by robertmartin8** [comment]
Hi,

Thanks for raising this issue, it's quite an interesting one. I suppose the reason why I didn't bother to avoid inversion of the omega matrix is that 100%-confidence views are not realistic. However, I am now of the opinion that this shouldn't be up to PyPortfolioOpt to dictate. 

I'm familiar with the Walters paper – actually the equations you provide are consistent with equations (25) and (30) from the paper. But it disagrees with Walters' equation in Appendix D! In fact, there is a typo in his Appendix D, which can be seen in the second last line where he messes up the sign when factorising. 

I'll have a go at rewriting the code this weekend. @schneiderfelipe do you have any thoughts before I proceed?

Best,
Robert

**Comment by jacobms** [comment]
Hi Robert
Awesome - thanks for the update.

I am confident that you have a far better foundation for knowing what is up/down in the papers and how to rework the implementing. I'm just happy that you are so quick to respond. Awesome!

All the best,
Jacob

**Comment by jacobms** [comment]
FYI stumbled upon this while reading the Walters paper: http://www.globalriskguard.com/resources/assetman/blach-litterman.pdf pp. 14-15 regarding BL with certain estimates including proof.

**Comment by schneiderfelipe** [comment]
> Hi,
> 
> Thanks for raising this issue, it's quite an interesting one. I suppose the reason why I didn't bother to avoid inversion of the omega matrix is that 100%-confidence views are not realistic. However, I am now of the opinion that this shouldn't be up to PyPortfolioOpt to dictate.
> 
> I'm familiar with the Walters paper – actually the equations you provide are consistent with equations (25) and (30) from the paper. But it disagrees with Walters' equation in Appendix D! In fact, there is a typo in his Appendix D, which can be seen in the second last line where he messes up the sign when factorising.
> 
> I'll have a go at rewriting the code this weekend. @schneiderfelipe do you have any thoughts before I proceed?
> 
> Best,
> Robert

I think this is perfect. The proposed implementation avoids all inversions, we only need to solve two linear systems with different right-hand sides:

```python3
tau_sigma_prior = tau * sigma_prior
B = tau_sigma_prior @ P.T
A = P @ B + Omega

pi_corr = B @ np.linalg.solve(A, Q - P @ pi_prior)
pi_posterior = pi_prior + pi_corr

sigma_corr = tau_sigma_prior - B @ np.linalg.solve(A, P @ tau_sigma_prior)
sigma_posterior = sigma_prior + sigma_corr
```

I also recall that another problem in the current implementation is that it's almost independent of tau (I have no idea why as I didn't get the time to look at it yet). But the proposed decomposition solves it perfectly, so I suggest that we make a test for varying, non-zero values of tau and ensure that things change (another obvious test would be for tau = 0, for which we get the prior exactly).

(@jacobms I like the suggestion of implementing Idzorek in the near future as well!)

**Comment by jacobms** [comment]
Great to hear 😊

Quick input (which was why I highlighted in my answer above): I don't have enough grasp whether whether there are consequences from following the approach of the Alternative Reference Model rather than the Canonical Reference Model (and Beyond Black-Litterman Reference Model) cf. Walters ("The Black-Litterman Model In Detail") i.e. whether mixing and matching is possible.

**Comment by schneiderfelipe** [comment]
> Great to hear blush
> 
> Quick input (which was why I highlighted in my answer above): I don't have enough grasp whether whether there are consequences from following the approach of the Alternative Reference Model rather than the Canonical Reference Model (and Beyond Black-Litterman Reference Model) cf. Walters ("The Black-Litterman Model In Detail") i.e. whether mixing and matching is possible.

I have no clue :see_no_evil:.
I'm not familiarised with those terms, sorry. What do you think @robertmartin8?

**Comment by robertmartin8** [comment]
@schneiderfelipe @jacobms 

There are a number of different issues being floated around here, so let me do my best to clarify.

### Reference models

The **reference model** is basically the problem setup for Black-Litterman. Assume that stocks are distributed normally, i.e r ~ N(mu, cov). The canonical reference model treats mu as a random variable centred around the prior, with another covariance matrix. The "Black-Litterman formula" is just a solution to the problem of best estimating mu and cov. 

By contrast, the alternative reference model does not treat mu as a random variable. Rather, it is just a quantity to be estimated. As Walters explains, this method removes the need for the parameter tau.

After specifying the reference model, we require an **estimation model** to fit parameters. PyPortfolioOpt's implements the Black Litterman "master formula", which applies to the canonical reference model. 

### Idzorek's method

Idzorek's method _seems_ that it would be a good idea, but it is actually not at all straightforward to implement. The API would need to be changed a little, as @schneiderfelipe points out, but the method itself is not trivial. I will add this to the feature backlog and happily accept contributions, but I don't have the bandwidth to implement it any time soon (particularly when there are more pressing things that need implementation).

### The fix

> I think this is perfect. The proposed implementation avoids all inversions, we only need to solve two linear systems with different right-hand sides:

Great, I will work on that.

>I also recall that another problem in the current implementation is that it's almost independent of tau (I have no idea why as I didn't get the time to look at it yet). But the proposed decomposition solves it perfectly, so I suggest that we make a test for varying, non-zero values of tau and ensure that things change (another obvious test would be for tau = 0, for which we get the prior exactly).

That's actually quite a major bug, but I will play around with it and add it to the test coverage.

**Comment by robertmartin8** [comment]
@schneiderfelipe 

This is what I have so far. If you have a moment, can you check that this is equivalent to your formulation? 

In particular, for `bl_cov` I use the shortcut that P tau Sigma is just the transpose of tau Sigma P^T (since sigma is symmetric).

I have checked that the code agrees with the previous implementation for non-pathological cases so I assume it is correct.

```python
    def bl_returns(self):
        tau_sigma_P = self.tau * self.cov_matrix @ self.P.T
        A = (self.P @ tau_sigma_P) + self.omega
        b = self.Q - self.P @ self.pi
        post_rets = self.pi + tau_sigma_P @ np.linalg.solve(A, b)
        return pd.Series(post_rets.flatten(), index=self.tickers)
```

```python
    def bl_cov(self):
        tau_sigma_P = self.tau * self.cov_matrix @ self.P.T
        A = (self.P @ tau_sigma_P) + self.omega
        b = tau_sigma_P.T
        M = self.tau * self.cov_matrix - tau_sigma_P @ np.linalg.solve(A, b)
        posterior_cov = self.cov_matrix + M
        return pd.DataFrame(posterior_cov, index=self.tickers, columns=self.tickers)
```

**Comment by schneiderfelipe** [comment]
@robertmartin8 Sorry for the late reply!

> This is what I have so far. If you have a moment, can you check that this is equivalent to your formulation?

It looks like the same math to me :)

> In particular, for `bl_cov` I use the shortcut that P tau Sigma is just the transpose of tau Sigma P^T (since sigma is symmetric).

Great catch!

As a suggestion, we could precompute both `tau_sigma_P` and `A` in your code. The following is computed twice:

> ```python
> tau_sigma_P = self.tau * self.cov_matrix @ self.P.T
> A = (self.P @ tau_sigma_P) + self.omega
> ```

So we could substitute for...

```python
if self._tau_sigma_P is None:
    self._tau_sigma_P = self.tau * self.cov_matrix @ self.P.T
if self._A is None:
    self._A = A = (self.P @ tau_sigma_P) + self.omega
```

...and use `self._tau_sigma_P` in place of `self._tau_sigma_P`, etc. throughout. What do you think?

**Comment by robertmartin8** [comment]
@schneiderfelipe 

I've made those changes.

>I also recall that another problem in the current implementation is that it's almost independent of tau (I have no idea why as I didn't get the time to look at it yet)

This actually took me a while to figure out. I wrote out the maths a few times and verified that the code agreed, so couldn't understand why the result was independent of tau. 

The answer, in fact, is here:

```python
    @staticmethod
    def _default_omega(cov_matrix, P, tau):
        """
        If the uncertainty matrix omega is not provided, we calculate using the method of
        He and Litterman (1999), such that the ratio omega/tau is proportional to the
        variance of the view portfolio.
        """
        return np.diag(np.diag(tau * P @ cov_matrix @ P.T))
```

Because the default method sets `omega` proportional to `tau`, we can take a factor of tau out of the following expression: `self._A = (self.P @ self._tau_sigma_P) + self.omega`

Thus when we compute the inverse of A (or equivalently solve the system), out comes a factor of 1/tau, which cancels with the other tau in this line:

`post_rets = self.pi + self._tau_sigma_P @ np.linalg.solve(self._A, b)`.

Hence as long as a user uses the default omega, `tau` is useless. Therefore this is actually expected behaviour! But I will add a note in the docs explaining it because it is not "expected". 

I will try to make a test with a different omega, to verify that changing tau does actually have some effect. 

Robert

**Comment by robertmartin8** [comment]
Ugh, annoyingly one of my tests doesn't pass for python 3.5. I reckon some of the BL code implicitly requires on dictionaries being ordered :( 

I'll see if it's a quick fix, otherwise I'll just drop support for 3.5 since this issue keeps happening
